Bluetooth: Mesh: Fix friend node establishment and message

LingaoM opened the pull request:
Fix friend node unable decrypt friend_cred encrypt poll message
when friendship not establishment.

Fix friend node cache segment message for given lpn with same
sequence number, cause to lpn replay attack.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
#30240

LingaoM commented:
@michal-narajowski FYI

trond-snekvik commented on subsys/bluetooth/mesh/transport.c:
```diff
@@ -727,7 +727,7 @@ static int sdu_recv(struct bt_mesh_net_rx *rx, uint8_t hdr, uint8_t aszmic,
 			.src = rx->ctx.addr,
 			.dst = rx->ctx.recv_dst,
-			.seq_num = rx->seq,
+			.seq_num = seg->seq_auth & 0xffffff,
```
This is not right. `seg` will only be defined when this function is called from `trans_seg`. What's the motivation for moving the net_rx->seq set here?

LingaoM replied:
The sequence will be modified, which will cause the sequence number of the segment message stored in the friend queue to be lower than that of the previously stored segment message, which will cause the LPN node to start a replay attack after receiving the message

trond-snekvik replied:
Okay, I see now. When we change the `net_rx->seq` in `trans_seg()`, it will be wrong when we enqueue it in the friend queue in the end of `bt_mesh_trans_recv()`. 

The change here doesn't work for single segment messages though, as they pass `seg=NULL` to the `sdu_recv()` function.
Suggest changing it to this, which maps nicely to the rule in the spec:

```suggestion
			.seq_num = seg ? (seg->seq_auth & 0xffffff) : rx->seq,
```

trond-snekvik commented on subsys/bluetooth/mesh/friend.c:
```diff
@@ -1250,9 +1250,10 @@ static void subnet_evt(struct bt_mesh_subnet *sub, enum bt_mesh_key_evt evt)
-				break;
 			}
-
+			break;
```
Should enqueue an update here as well though, right?

LingaoM replied:
I didn't see that the specification required to add the friend update. After receiving the net_key_update message. But adding this friend update does not affect the state of the LPN.

trond-snekvik replied:
You're right, the KR flag only changes in KR phase 2. I'll make a function for this rule in a new PR, this logic is repeated in various places. This works for this PR, though.

trond-snekvik reviewed (changes requested):
> Fix friend node cache segment message for given lpn with same sequence number, cause to lpn replay attack.

I saw the description in the issue, but it's a bit unclear to me how this patch is fixing this. Could you please explain?

See additional comments

jhedberg reviewed (changes requested):
@LingaoM do I understand correctly that there are two independent fixes here? At least it sounds like it based on the commit message. In that case you should really split this up into two separate commits.

Also, did you intend to put a "Fixes: ..." to the description, since you've reference one issue in the comments?

LingaoM committed:
Bluetooth: Mesh: Fix friend node cache message

Friend node cache segment message for lpn, message with
sequence number can't change, since could cause lpn replay
attack.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Fix Friend node handle Key update

When friend node key refresh phase to stage_1, there
are need to add friend update message to friend queue,
but, when friend KR phase to stage_2, must add friend
update to queue.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Fix Friend node estalished

Since first poll request send by lpn use friend security
credentials, so, friend nodes should be able to decrypt with
friend security, even if they have not yet established a friendship.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>
